### PR TITLE
Fix ProfileTransfer cleanup re-entry crash

### LINKED ISSUE #5612: Profile transfer failure cleanup can re-enter and crash
## Problem

A profile import/upload can crash AetherSDR when a timeout or another failure starts cleanup while an incomplete upload socket is still connected.

## Root cause

`ProfileTransfer::cleanup()` calls `m_socket->abort()` while the socket remains connected to `onUploadDisconnected()` and remains stored in `m_socket`. `abort()` synchronously emits `disconnected`; when the upload is incomplete, `onUploadDisconnected()` calls `fail()` and re-enters `cleanup()`. The inner cleanup nulls `m_socket`, then the outer cleanup resumes and calls `m_socket->deleteLater()` through the null member.

The sibling transfer implementations in `FirmwareUploader` and `WaveformInstaller` already avoid this pattern by detaching the member and disconnecting handlers before aborting a local pointer.

## Reproduction evidence

An ARM64 AddressSanitizer harness using the production `ProfileTransfer.cpp` reproduced the fault with a connected local descriptor pair and an incomplete upload:

```text
Before fail: socket state QAbstractSocket::ConnectedState
ProfileTransfer failed: "Injected transfer timeout"
ProfileTransfer failed: "Radio closed the upload connection before the transfer completed."
AddressSanitizer: SEGV on unknown address 0x000000000008
    QObject::deleteLater()
    AetherSDR::ProfileTransfer::cleanup()
    AetherSDR::ProfileTransfer::fail(QString const&)
```

The descriptor pair establishes `ConnectedState` only; it does not run a listener, simulated radio, or network connection.

## Expected behavior

Cleanup must be idempotent. Failure, timeout, or destruction with an active incomplete transfer should emit at most one terminal outcome and must not crash.

## Proposed scope

- Detach the active socket member and disconnect its handlers before aborting/deleting it.
- Ensure terminal cleanup cannot recursively emit a second failure.
- Add a focused, mutation-checked regression test for failure cleanup with an active incomplete socket.

Found during the 2026-09-12 codebase bug review at `d8fe821329119ce9eca7d2d5268e7a8765ce2749`.

### PR DESCRIPTION
**Pull request by rfoust**
## Summary

Fixes #5612. A timeout or other failure during an incomplete profile upload could crash because `ProfileTransfer::cleanup()` aborted its active socket while the socket was still stored in `m_socket` and still connected to `onUploadDisconnected()`. The synchronous disconnect re-entered `fail()` and `cleanup()`, cleared the member, and left the outer cleanup dereferencing a null pointer.

This change centralizes socket disposal so the member is cleared and its callbacks are disconnected before aborting or disconnecting. The same helper now covers failure cleanup, fallback socket replacement, and successful upload completion. Cleanup also enters its inactive terminal state before tearing down any QObject, preventing unrelated synchronous callbacks from starting a second terminal outcome.

The regression test uses an inert `QTcpSocket` state-machine object: it opens no listener, descriptor, connection, or simulated firmware peer. It injects an independent synchronous callback during `abort()` and verifies that cleanup emits exactly one terminal failure. A narrow friend accessor reaches the private lifecycle seam without preprocessor access rewriting.

## Constitution principle honored

Principle XI — Fixes Are Demonstrated. The regression test fails with the original abort-before-detach ordering and passes with the fix restored.

## Test plan

- [x] Native ARM64 configure completed with RADE enabled
- [x] `profile_transfer_test` passes
- [x] `profile_transfer_cleanup_test` passes
- [x] `python3 tools/check_test_registration.py --strict` passes
- [x] Mutation check: restoring abort-before-detach produces the original double failure and segmentation fault
- [x] Guard mutation check: moving terminal-state clearing back after QObject teardown produces two failure emissions and fails the regression test
- [x] Existing CI tests pass on Linux, macOS, and Windows
- [x] Real-radio verification is not applicable to this deterministic QObject/socket lifecycle defect

## Agent automation bridge

The bridge was launched against the built application with isolated settings and transmit disabled; it reported AetherSDR 26.9.2, `txAllowed: false`, and exited cleanly through the bridge. Its 74-verb inventory has no profile-transfer lifecycle or socket-fault injection surface, so it cannot exercise this internal failure path. The operator explicitly approved publication using the production-linked, mutation-checked regression test as the behavioral proof.

## Checklist

- [x] Commit is SSH-signed (`docs/COMMIT-SIGNING.md`)
- [x] No new flat-key `AppSettings` calls
- [x] Code is clean-room
- [x] No meter UI changes
- [x] No user-facing documentation or `CHANGELOG.md` change is needed
- [x] No security-sensitive change or GHSA applies

Generated with OpenAI Codex.

---

## Review follow-up (49b592b3, a1b78897)

Review found that the original regression test proved the terminal-state guard
and the `connectUploadSocket()` detach, but three parts of the fix were
invisible to it — reverting them left the test green. Addressed:

- **`cleanup()`'s detach-before-abort is now asserted directly.** The teardown
  observer records whether the transfer had already released the socket at the
  instant it reached `UnconnectedState`, which is the ordering rule itself
  rather than a symptom of breaking it. Previously the `m_busy` hoist masked a
  missing detach at this site.
- **The successful-completion path (`destroySocket(false)`) now has coverage.**
  A third case drives a complete upload closed by the radio. This is the one
  disposal site where the transfer is still busy, so the terminal-state guard
  cannot stand in for detaching the handlers. Re-entry is pinned by counting
  the completion branch's own `progress` emission, not raw `disconnected()`
  emissions — Qt produces a variable number of those as a socket winds down.
- **The download listener gets the same detach-then-act treatment**, via a new
  `destroyServer()`. `QTcpServer::close()` does not emit synchronously, so this
  is not a live crash; but a `newConnection` queued before cleanup would reach
  `onDownloadConnection()` with `m_server` already null, stopped today only by
  the busy guard. A fourth case pins it with an inert `QTcpServer` that never
  calls `listen()` — it binds no port and owns no descriptor, so the test stays
  socket-free.

### Mutation matrix

Each reverted independently, rebuilt, and run:

| Reverted | Before | After |
|---|---|---|
| `cleanup()` detach before abort | exit 0 — **not caught** | **exit 1** |
| Terminal-state guard placement | exit 1 | exit 1 |
| `connectUploadSocket()` detach | exit 139 (SEGV) | exit 139 (SEGV) |
| Completion path shared helper | exit 0 — **not caught** | **exit 1** |
| `destroyServer()` detach before close | n/a | **exit 1** |

Unmutated: passes. Full app target builds clean;
`check_test_registration.py --strict` passes.

### Disclosed behavior change

Routing successful completion through the shared helper means it now calls
`disconnectFromHost()` where the old code only called `deleteLater()`. This is
behavior-neutral: `onUploadDisconnected()` is reachable only from
`QTcpSocket::disconnected`, so the socket is already `UnconnectedState` and the
call returns immediately. It is kept because it matches the sibling
`FirmwareUploader`/`WaveformInstaller` helpers this fix is a port of.

### CODE REVIEW

File: src/core/ProfileTransfer.cpp [hunk 1]
```diff
@@ -217,15 +217,16 @@ void ProfileTransfer::finish(QString path)
...
     m_timeout->stop();
     m_idleTimer->stop();
     m_overallTimer->stop();
 
-    if (m_socket) {
-        m_socket->abort();
-        m_socket->deleteLater();
-        m_socket = nullptr;
-    }
+    destroySocket(true);
```
Comment (ten9876) [inline src/core/ProfileTransfer.cpp:229]: **Nit, non-blocking** — this line's contribution is not pinned by the new test.

I reverted exactly this call back to the pre-PR inline block, leaving the terminal-state guard above intact:

```cpp
if (m_socket) {
    m_socket->abort();
    m_socket->deleteLater();
    m_socket = nullptr;
}
```

and `profile_transfer_cleanup_test` **passed (exit 0)**. It passes because `m_busy = false` at line 222 already makes `onUploadDisconnected`'s `if (m_cancelled || !m_busy) return;` trip, so the re-entrant `fail()` never happens and the null deref never materializes — the guard alone is sufficient at *this* site.

That is not a correctness problem: reverting both halves (the true pre-PR state) does segfault, so the actual #5612 regression is caught, and the same revert at `connectUploadSocket` (line 417) segfaults on its own. The gap is durability — someone later trimming the detach here as "redundant, the guard covers it" would get a green test. A cheap way to pin it, if you want it, is a third block that drives `cleanup()` with the guard's protection deliberately unavailable, or simply asserts the socket has no remaining receivers on the transfer before `abort()` is allowed to run:

```suggestion
    destroySocket(true); // detach-then-abort: see profile_transfer_cleanup_test
```

(The suggestion is just the comment marker — the real ask is the extra assertion, which is your call.)

File: src/core/ProfileTransfer.cpp [hunk 2]
```diff
@@ -500,10 +512,7 @@ void ProfileTransfer::onUploadDisconnected()
         return;
     }
 
-    if (m_socket) {
-        m_socket->deleteLater();
-        m_socket = nullptr;
-    }
+    destroySocket(false);
```
Comment (ten9876) [inline src/core/ProfileTransfer.cpp:529]: **Nit, non-blocking** — two small things about the success-completion site.

1. **No coverage.** Reverting this line to the original `if (m_socket) { m_socket->deleteLater(); m_socket = nullptr; }` leaves `profile_transfer_cleanup_test` passing (exit 0). The body lists "successful upload completion" as one of the three sites the helper now covers, and it does — but nothing pins it. This is the one site where the `m_busy` guard does *not* help, since `m_busy` is still `true` here, so the detach is doing real work.

2. **Small undisclosed behavior change.** The old code only called `deleteLater()`; this now also calls `disconnectFromHost()`. I convinced myself it is a no-op in practice — `onUploadDisconnected` is only reachable from `QTcpSocket::disconnected`, so the socket is already `UnconnectedState` and `disconnectFromHost()` returns immediately. It also matches the sibling `FirmwareUploader`/`WaveformInstaller` shape, so I would keep it. Just worth a line in the PR body.

File: tests/profile_transfer_cleanup_test.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                     "socket enters an incomplete connection state before cleanup");
+
+        AetherSDR::ProfileTransferTestAccess::fail(
+            transfer, QStringLiteral("forced cleanup failure"));
+
+        ok &= expect(abortDisconnects == 1,
+                     "failure cleanup synchronously aborts the incomplete socket");
+        ok &= expect(!transfer.isBusy(), "failure cleanup leaves the transfer idle");
+        ok &= expect(AetherSDR::ProfileTransferTestAccess::socket(transfer) == nullptr,
+                     "failure cleanup clears the active socket");
```
Comment (ten9876) [inline tests/profile_transfer_cleanup_test.cpp:164]: For the record, the assertions in this block do hold up under mutation — I am not questioning the test's value, only its reach.

What it provably pins (verified by building each mutation and running it):

- moving `m_busy`/`m_cancelled` back below the QObject teardown → **exit 1**, `FAIL: cleanup cannot emit a re-entrant second failure`
- reverting `connectUploadSocket` to abort-before-detach → **exit 139 (SEGV)**, caught by the second block
- reverting both halves at the `cleanup()` site → **exit 139 (SEGV)**

What it does not pin is the `cleanup()`-site detach in isolation (see the comment on `ProfileTransfer.cpp:229`). `failure cleanup clears the active socket` on this line passes either way, because the inline version also nulls `m_socket` — it just does so after `abort()` rather than before, which is the whole bug and is invisible once the guard is in place.

One robustness note while you are here: the second block calls the real `connectUploadSocket`, which schedules `QTimer::singleShot(200, this, …)` that would dereference `m_model` — `nullptr` in this test. It cannot fire today (no event loop, and `transfer` dies at scope exit), so the test is correct as written. But if anyone later adds a `QCoreApplication::processEvents()` or a `QTest::qWait` to this file, that timer becomes a null-deref crash. A one-line comment saying the test deliberately runs without an event loop would save the next person the investigation.

**Review by ten9876 — COMMENTED** [review]
## Issue fit

This solves #5612. The issue asks for three things — detach the socket member before aborting it, make terminal cleanup unable to emit a second outcome, and add a mutation-checked regression test — and the diff delivers all three. `destroySocket(bool)` is a verbatim port of the existing `FirmwareUploader::destroySocket` / `WaveformInstaller::destroySocket` helpers (same name, same signature, same body), which is exactly the precedent the issue thread pointed at, so this lands as consistency rather than a new invention. It also covers the second site the issue's triage comment flagged and the original report did not (`connectUploadSocket`, `ProfileTransfer.cpp:417`), and as a bonus it protects the *download* socket too — `onDownloadConnection` wires `readyRead`/`disconnected`/`errorOccurred` to `this`, so the same `QObject::disconnect(socket, nullptr, this, nullptr)` severs those as well. After the change there is no raw `m_socket->abort()` / `m_socket->deleteLater()` left anywhere in the file; teardown is fully centralized.

No blockers. One nit below about what the regression test actually pins, backed by a mutation matrix I ran locally.

## Scope

| File | Change | Claimed in title/body? | Verdict |
|---|---|---|---|
| `src/core/ProfileTransfer.cpp` | Centralized `destroySocket(bool)`; terminal state set at top of `cleanup()`; three call sites converted | Yes | In scope |
| `src/core/ProfileTransfer.h` | `destroySocket` decl + `friend class ProfileTransferTestAccess` | Yes | In scope |
| `tests/profile_transfer_cleanup_test.cpp` | New regression test | Yes | In scope |
| `tests/tests.cmake` | Registers the test | Yes | In scope |

Everything in the diff is explained by the issue. One commit, authored the same day the issue was filed, message references #5612 — no stale or cherry-picked outliers. No new protocol verb, settings key, capability field, CLI flag or exported API; no `CHANGELOG.md` entry (correct — it is release-prep only); no UI, default-value or a11y change, so there is no preference-smuggling surface here at all. The `-` side of the diff removes only the two inline teardown blocks that `destroySocket` replaces — no guard, early return or issue-citing comment was deleted. `friend class …TestAccess` is the established seam in this tree (29 uses, e.g. `RadioSetupDialogTestAccess`, `DigitalVoiceWaveformProcessTestAccess`), not a new pattern.

Body checklist holds up: the commit is SSH-signed and GitHub reports `verified=true reason=valid`; no flat-key `AppSettings` calls; linking `aethercore` is the sanctioned CMake path (162 existing targets do it), so the `AETHER_SETTINGS_CONSUMERS` obligation — which applies to targets compiling `${AETHER_SETTINGS_SOURCES}` *directly* — does not attach.

## Test-layer boundary (#5232 / #5254)

Checked before building anything, since the new test names `QTcpSocket`. It is **not** a socket-owning test and needs no operator gate: `IncompleteSocket` forces its state with `setSocketState(ConnectingState)` and never opens a descriptor, binds, listens or connects, and the one production path that would dial out (`connectUploadSocket`) defers `connectToHost` behind a 200 ms `QTimer::singleShot` that cannot fire — the test has no event loop and destroys the transfer at scope exit. No fake firmware peer. Correct layer for a deterministic QObject-lifetime defect.

## Nits (non-blocking)

1. **The test pins the re-entrancy guard, but not the detach ordering at two of the three call sites.** See the inline comment on `ProfileTransfer.cpp:229`. Mutation matrix, built and run locally at `2adb234`:

   | Mutation | Result |
   |---|---|
   | B — terminal state moved back below the QObject teardown (detach kept) | **exit 1** — `FAIL: cleanup cannot emit a re-entrant second failure` ✅ |
   | C — `connectUploadSocket:417` reverted to inline abort-before-detach | **exit 139 (SEGV)** ✅ |
   | AB — both reverted at the cleanup site (true pre-PR state) | **exit 139 (SEGV)** ✅ |
   | A — `cleanup():229` reverted to inline abort-before-detach, guard kept | **exit 0 — passes** ⚠️ |
   | D — `onUploadDisconnected():515` reverted to plain `deleteLater` | **exit 0 — passes** ⚠️ |

   So the regression genuinely described in #5612 *is* caught (AB segfaults), and the body's guard-mutation claim is exactly right. But the body's first claim — "restoring abort-before-detach produces the original double failure and segmentation fault" — only holds when the guard is reverted too. With the guard in place, removing the detach at `cleanup()` is silent. Worth knowing, because a future refactor that trims `destroySocket`'s detach as "redundant" would pass CI.

2. **`destroySocket(false)` on the success path (`:515`) has no coverage and is a small undisclosed behavior change** — see the inline comment there. Low risk, and I could not construct a failure from it; noted for completeness.

3. `m_server` and `m_saveFile` in `cleanup()` still use the detach-last shape the socket just moved away from. Neither emits synchronously on `close()`/`cancelWriting()`, and the new `m_busy` guard blocks re-entry regardless, so this is genuinely fine today — just noting the asymmetry now that the socket sets a better local precedent.

## What I tried to break, and what held

- **Built the PR head and ran both targets** in a scratch worktree (Debug, Ninja): `profile_transfer_test` and `profile_transfer_cleanup_test` both pass. Then ran the five-way mutation matrix above — this is where the finding came from; every other mutation behaved as the body claims.
- **`cancel()` under the moved terminal state.** `cancel()` sets `m_cancelled = true`, and `cleanup()` now clears it at the *top* rather than the bottom. Chased whether that changes the emitted outcome: it does not — `cancel()` snapshots `m_operation` before `cleanup()` and emits `failed` after, and `fail()`'s `!m_busy && !m_cancelled` guard leaves both false either way, so a cancel still produces exactly one terminal signal with the same text. Observers of `isBusy()` also see no change, since the old code already cleared `m_busy` inside `cleanup()` and therefore before the emit.
- **Destructor path.** `~ProfileTransfer()` calls `cleanup()` when busy, and `destroySocket` calls `deleteLater()` on a socket that is a child of the object being destroyed. Qt drops the pending deferred-delete when the parent deletes the child, so no double free; the test's first block exercises exactly this teardown and is clean.
- **The download side**, which the body does not mention. `onDownloadConnection` takes `m_socket` from `m_server->nextPendingConnection()` (parented to the server, not to the transfer) and wires three handlers to `this`. Confirmed `destroySocket` severs all three and that the socket/server `deleteLater` ordering is safe. The fix helps here too.
- **The bridge claim.** Verified rather than taken on faith: `AutomationServer.cpp` contains **zero** references to `ProfileTransfer`, so the body is right that no automation verb can reach this lifecycle. That is why I did not drive the app for this one — there is no reachable surface, not because I skipped it. Findings here are from build + mutation execution, not from reading alone.
- **Leftovers.** Grepped for any surviving `m_socket->abort()` / `m_socket->deleteLater()` / `m_socket = nullptr` outside the helper: the only hit is line 260 inside `destroySocket` itself.

CI is green on all five checks; note it ran against merge base `623a97a1` while `main` is now `a3d1971a` (#5618, #5521 landed since). Those touch model verb gating and the HL2 SWR detector, nothing this diff reaches, and GitHub still reports the PR `MERGEABLE`.

File: src/core/ProfileTransfer.cpp (hunk 1 above)
Comment (ten9876) [reply src/core/ProfileTransfer.cpp:229]: Fixed in 49b592b3. The test now asserts the ordering directly rather than its consequence: the teardown observer records whether the transfer had already released the socket at the instant it reached UnconnectedState, so `failure cleanup detaches the socket before aborting it` fails when this line is reverted to the inline block — mutation re-run gives **exit 1** where it previously gave exit 0. I dropped the marker-comment suggestion; the assertion is the durable version of it.

File: src/core/ProfileTransfer.cpp (hunk 2 above)
Comment (ten9876) [reply src/core/ProfileTransfer.cpp:529]: Both points addressed in 49b592b3.

1. **Coverage** — a third test case now drives the completion path (complete upload, socket in ConnectedState, radio closes the connection). Reverting this line to the bare `deleteLater()` now fails two assertions: `completion disposes the socket through the shared helper` and `completion detaches the socket before disconnecting it` (**exit 1**, previously exit 0). Re-entry is pinned by counting the completion branch's own progress emission rather than raw `disconnected()` emissions, which Qt produces a variable number of as the socket winds down.

2. **The `disconnectFromHost()` addition** is now stated in the PR body as an intentional, behavior-neutral consequence of routing through the shared helper.

File: tests/profile_transfer_cleanup_test.cpp (hunk 3 above)
Comment (ten9876) [reply tests/profile_transfer_cleanup_test.cpp:164]: Addressed in 49b592b3 and a1b78897. The reach gap is closed — every disposal site is now independently pinned (full matrix in the PR body). The event-loop trap is now documented twice: a file-level note above `main()` explaining that nothing is pumped on purpose, and an inline note at the `connectUploadSocket()` call site, which is where the null-`m_model` timer actually lives.

File: src/core/ProfileTransfer.cpp [hunk 4]
```diff
@@ -217,15 +217,16 @@ void ProfileTransfer::finish(QString path)
 
 void ProfileTransfer::cleanup()
 {
+    // Enter the terminal state before tearing down QObjects. Their methods can
+    // emit synchronously, and no callback may start a second terminal outcome.
+    m_busy = false;
```
[the remark this answers is not in the document]
Comment (ten9876) [reply src/core/ProfileTransfer.cpp:222]: Agreed, and confirmed independently — I reverted the hoist on its own (keeping `destroySocket()`) and block 1 fails with `cleanup cannot emit a re-entrant second failure`. Your reading of `cancel()` matches mine: `m_operation` is snapshotted before `cleanup()` and the emit follows it, so the net terminal state and message are unchanged. No action needed.

File: src/core/ProfileTransfer.cpp [hunk 5]
```diff
@@ -243,14 +244,29 @@ void ProfileTransfer::cleanup()
...
+    }
+
+    QTcpSocket* socket = m_socket;
+    m_socket = nullptr;
+    QObject::disconnect(socket, nullptr, this, nullptr);
+    if (abortConnection) {
+        socket->abort();
+    } else {
+        socket->disconnectFromHost();
+    }
```
[the remark this answers is not in the document]
Comment (ten9876) [reply src/core/ProfileTransfer.cpp:262]: Confirming the citation-health note: both siblings do exist at `src/core/FirmwareUploader.cpp:655` and `src/core/WaveformInstaller.cpp:479`, each with this exact helper under the same name and signature. The checker was pointed at the wrong tree; the precedent is real.

Your two checks match what I found, and the `abortConnection = false` branch now has explicit coverage as of 49b592b3 — it was the one disposal site nothing observed. `destroyServer()` in the same commit gives the download listener the same detach-then-act treatment, pinned by an inert `QTcpServer` that never calls `listen()`.

File: tests/profile_transfer_cleanup_test.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        AetherSDR::ProfileTransfer transfer(nullptr);
+        int failures = 0;
+        int abortDisconnects = 0;
+        QObject::connect(&transfer, &AetherSDR::ProfileTransfer::failed,
+                         [&failures](AetherSDR::ProfileTransfer::Operation, const QString&) {
+                             ++failures;
+                         });
+
+        IncompleteSocket* oldSocket = prepareIncompleteUpload(transfer, abortDisconnects);
+        AetherSDR::ProfileTransferTestAccess::connectUploadSocket(transfer, 42607);
```
[the remark this answers is not in the document]
Comment (ten9876) [reply tests/profile_transfer_cleanup_test.cpp:183]: Applied in a1b78897 — your suggestion verbatim at the call site, plus a file-level note above `main()` so the constraint is visible before anyone reaches this line. Good catch on the second-order consequence: it isn't just a null deref, it would also turn a socket-free test into one that dials out, which is the part that would have slipped past the test-layer boundary review.

**Review by ten9876 — APPROVED** [review]
## Approving

All nits from the previous review are fixed and pushed (`49b592b3`, `a1b78897`); all six review threads are resolved.

The underlying fix was correct from the start — `destroySocket(bool)` is a verbatim port of the helper already carried by `FirmwareUploader` and `WaveformInstaller`, applied to all three disposal sites, and it also covers the download socket, which the original body did not claim. The only real weakness was that the regression test proved less than the body said it did. That is now closed.

**What changed since the last review**

- `cleanup()`'s detach-before-abort is asserted directly, by recording whether the transfer had already released the socket at the instant it reached `UnconnectedState`. Previously the `m_busy` hoist masked a missing detach here, so this line could be deleted with CI green.
- The successful-completion path (`destroySocket(false)`) has coverage for the first time. It is the one disposal site where the transfer is still busy, so the terminal-state guard genuinely cannot substitute for detaching the handlers.
- The download listener now goes through a matching `destroyServer()`. Not a live crash — `QTcpServer::close()` does not emit synchronously — but a `newConnection` queued before cleanup would reach `onDownloadConnection()` with `m_server` already null, stopped today only by the busy guard. Pinned by an inert `QTcpServer` that never calls `listen()`, so the test stays socket-free.
- The `disconnectFromHost()` addition on the completion path is now disclosed in the body as behavior-neutral, with the reasoning.

**Mutation matrix** — each reverted independently, rebuilt, and run:

| Reverted | Before | After |
|---|---|---|
| `cleanup()` detach before abort | exit 0 — not caught | **exit 1** |
| Terminal-state guard placement | exit 1 | exit 1 |
| `connectUploadSocket()` detach | exit 139 (SEGV) | exit 139 (SEGV) |
| Completion path shared helper | exit 0 — not caught | **exit 1** |
| `destroyServer()` detach before close | n/a | **exit 1** |

Every line of the fix is now load-bearing for the test. Unmutated passes, the full `AetherSDR` target builds clean, and `check_test_registration.py --strict` passes.

Still socket-free and at the right layer for a deterministic QObject-lifetime defect: no target binds, listens, connects, or stands up a firmware peer. Scope remains exactly the linked issue — no protocol surface, no settings key, no UI or default change, no `CHANGELOG.md` entry.

Thanks @rfoust — the fix itself needed no correction, and matching the existing sibling helper rather than inventing a new shape is what made the rest of this straightforward to verify.
